src/Compilers/CSharp/Portable/FlowAnalysis/PreciseAbstractFlowPass.cs

```diff
@@ -2010,6 +1972,47 @@ protected void AfterRightChildOfBinaryLogicalOperatorHasBeenVisited(BoundExpress
             }
         }
 
+        protected static BoundExpression GetBinaryLogicalOperatorLeft(BoundExpression expr)
+        {
+            switch (expr.Kind)
+            {
+                case BoundKind.BinaryOperator:
+                    {
+                        var binOp = (BoundBinaryOperator)expr;
+                        if (binOp.OperatorKind.IsLogical())
+                        {
+                            return binOp.Left;
+                        }
+                    }
+                    break;
+                case BoundKind.UserDefinedConditionalLogicalOperator:
+                    return ((BoundUserDefinedConditionalLogicalOperator)expr).Left;
+            }
+            return null;
+        }
+
+        protected static BoundExpression GetBinaryLogicalOperatorRight(BoundExpression expr, out BinaryOperatorKind kind)
+        {
+            switch (expr.Kind)
+            {
+                case BoundKind.BinaryOperator:
+                    {
+                        var binOp = (BoundBinaryOperator)expr;
+                        kind = binOp.OperatorKind;
+                        Debug.Assert(kind.IsLogical());
+                        return binOp.Right;
+                    }
+                case BoundKind.UserDefinedConditionalLogicalOperator:
+                    {
+                        var binOp = (BoundUserDefinedConditionalLogicalOperator)expr;
+                        kind = binOp.OperatorKind;
+                        return binOp.Right;
+                    }
+                default:
+                    throw ExceptionUtilities.UnexpectedValue(expr.Kind);
+            }
+        }
```

Comment:
I like the refactoring done in this file, but it seems unrelated to the PR. 
Would it be better to make this change in dev15.7.x as a separate PR, as you recently did with another cleanup?

Reply:
The refactoring is needed for this PR so that `NullableWalker` has access to the stack of binary operators. I'll consider porting the changes to `PreciseAbstractFlowPass` to dev15.7.x though.

---
In reply to: [164230694](https://github.com/dotnet/roslyn/pull/23981#discussion_r164230694) [](ancestors = 164230694)